### PR TITLE
Merging models

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/912.

The general design is as in the above mentioned issue. 2 things that are kept in mind when merging model B into model A:
- B itself should not be modified, so a copy of it is made and this copy is modified. This modified B is then passed on to merge methods on node and table level;
- Only the merge method on the model level makes modifications, not the ones on node and table level. Only the table merge method does asserts for node ID conflicts (so allocation network IDs are not checked, as these are not unique. Also node/edge geometries are not checked).

### CODE REVIEW

**Comment by SouthEndMusic** [comment]
@evetion mypy still complains about a few things, I don't know what the best way would be to fix them.

File: python/ribasim/tests/test_model.py [hunk 1]
```diff
@@ -143,3 +144,19 @@ def test_write_adds_fid_in_tables(basic, tmp_path):
...
+
+
+def test_model_merging(basic, subnetwork, tmp_path):
+    model = deepcopy(basic)
+    model_added = deepcopy(subnetwork)
+    model.merge_model(model_added)
+    model.merge_model(model_added)
+    assert (model.network.node.df.index == range(1, 44)).all()
+    assert model.max_allocation_network_id() == 2
+    for node_type, node_added in model_added.nodes().items():
```
Comment (SouthEndMusic) [inline python/ribasim/tests/test_model.py:156]: Regarding this loop: I tried comparing models directly, e.g.
```python
assert model_added == subnetwork
```

but I kept running into 'maximum recursion depth reached' errors, and I could not find out why; there is no loop in the model object as far as I could find.

File: python/ribasim_testmodels/ribasim_testmodels/allocation.py [hunk 2]
```diff
@@ -1065,3 +1071,123 @@ def allocation_example_model():
     )
 
     return model
+
+
+def main_network_with_subnetworks_model():
```
Comment (SouthEndMusic) [inline python/ribasim_testmodels/ribasim_testmodels/allocation.py:1076]: I'll pick up this model in https://github.com/Deltares/Ribasim/pull/851 for testing.

File: python/ribasim/ribasim/geometry/node.py [hunk 3]
```diff
@@ -59,6 +63,34 @@ def node_ids_and_types(*nodes):
 
         return node_id, node_type
 
+    def translate_spacially(
```
Comment (evetion) [inline python/ribasim/ribasim/geometry/node.py:66]: Both functions here are identical to the one in edge.py, try to generalize these.

File: python/ribasim/ribasim/input_base.py [hunk 4]
```diff
@@ -202,6 +216,44 @@ def node_ids(self) -> set[int]:
 
         return node_ids
 
+    def offset_node_ids(self, offset_node_id: int) -> "TableModel[TableT]":
```
Comment (evetion) [inline python/ribasim/ribasim/input_base.py:219]: This does also create a copy, which could be reflected in the name. Maybe we want to have a mutable version as well?

File: python/ribasim/ribasim/input_base.py [hunk 5]
```diff
@@ -202,6 +216,44 @@ def node_ids(self) -> set[int]:
...
+                "node_id",
+                "from_node_id",
+                "to_node_id",
+                "listen_feature_id",
+            ]:
+                if hasattr(df, name_column):
+                    df[name_column] += offset_node_id
+        return copy
+
+    def merge_table(
```
Comment (evetion) [inline python/ribasim/ribasim/input_base.py:235]: I'd expected this function to also fix any index/id clashes by calling some form of offset?

File: python/ribasim/ribasim/input_base.py [hunk 6]
```diff
@@ -421,6 +473,61 @@ def node_ids_and_types(self) -> tuple[list[int], list[str]]:
...
+        """Add the same offset to all node IDs in all underlying tables."""
+        node_copy = deepcopy(self)
+        for field in node_copy.fields():
+            attr = getattr(node_copy, field)
+            if isinstance(attr, TableModel):
+                table = attr
+                setattr(
+                    node_copy,
+                    field,
+                    table.offset_node_ids(offset_node_id),
```
Comment (evetion) [inline python/ribasim/ribasim/input_base.py:486]: This will again make a copy of the already copied table (by deepcopy)!

File: python/ribasim/ribasim/model.py [hunk 7]
```diff
@@ -58,6 +61,75 @@ def n_nodes(self):
 
         return n
 
+    def translate_spacially(
```
Comment (evetion) [inline python/ribasim/ribasim/model.py:64]: Time to do a search-replace on spacially ;)

File: python/ribasim/ribasim/model.py [hunk 8]
```diff
@@ -376,6 +448,68 @@ def reset_contextvar(self) -> "Model":
...
+    def max_node_id(self) -> int:
+        return self.network.node.df.index.max()
+
+    def max_allocation_network_id(self) -> int:
+        m = self.network.node.df.allocation_network_id.max()
+        if pd.isna(m):
+            m = 0
+        return m
+
+    def merge_model(
```
Comment (evetion) [inline python/ribasim/ribasim/model.py:460]: nit: Your function names are sometimes a bit verbose. It's not wrong, but you might want to think about it. This would result in `modela.merge_model(modelb)`. Purely `merge` would thus also work.

File: python/ribasim/ribasim/model.py [hunk 9]
```diff
@@ -376,6 +448,68 @@ def reset_contextvar(self) -> "Model":
...
+        self,
+        model_added: "Model",
+        offset_node_id: int | None = None,
+        offset_allocation_network_id: int | None = None,
+        offset_spacial: tuple[float, float] = (0.0, 0.0),
+        inplace: bool = True,
+    ):
+        """
+        Merge copies of the nodes and edges of an added model into this model.
+        The added model is not modified, but the following modificadions are made to the added data:
```
Comment (evetion) [inline python/ribasim/ribasim/model.py:470]:
```suggestion
        The added model is not modified, but the following modifications are made to the added data:
```

A spellcheck in your IDE might help you 😄

File: python/ribasim/ribasim/model.py [hunk 10]
```diff
@@ -376,6 +448,68 @@ def reset_contextvar(self) -> "Model":
...
+        )
+        min_offset_node_id = model.max_node_id()
+        min_offset_allocation_network_id = model.max_allocation_network_id()
+
+        if offset_allocation_network_id is None:
+            offset_allocation_network_id = min_offset_allocation_network_id
+        else:
+            assert (
+                offset_allocation_network_id >= min_offset_allocation_network_id
+            ), f"The allocation network ID offset must be at least the maximum allocation network ID of the main model ({min_offset_allocation_network_id}) to avoid conflicts."
```
Comment (evetion) [inline python/ribasim/ribasim/model.py:492]: If you always check this, you might as well set the default argument to 0 instead of None.

File: python/ribasim/ribasim/model.py [hunk 11]
```diff
@@ -443,11 +577,22 @@ def plot(self, ax=None) -> Any:
         if ax is None:
             _, ax = plt.subplots()
             ax.axis("off")
+
         self.network.edge.plot(ax=ax, zorder=2)
         self.plot_control_listen(ax)
         self.network.node.plot(ax=ax, zorder=3)
 
-        ax.legend(loc="lower left", bbox_to_anchor=(1, 0.5))
+        handles, labels = ax.get_legend_handles_labels()
```
Comment (evetion) [inline python/ribasim/ribasim/model.py:585]: This should ideally not be part of the PR (but will ignore it here)?

File: python/ribasim/tests/test_model.py (hunk 1 above)
Comment (evetion) [reply python/ribasim/tests/test_model.py:158]: It might be that a comparison somewhere (especially given that you implement `__eq__` yourself) is calling itself again.

File: python/ribasim/tests/test_model.py [hunk 12]
```diff
@@ -143,3 +145,20 @@ def test_write_adds_fid_in_tables(basic, tmp_path):
         assert "fid" in df.columns
         fids = df.get("fid")
         assert fids.equals(pd.Series(range(1, len(fids) + 1)))
+
+
+def test_model_merging(basic, subnetwork, tmp_path):
+    model = deepcopy(basic)
+    model_added = deepcopy(subnetwork)
+    model.merge_model(model_added)
+    model.merge_model(model_added)
```
Comment (evetion) [inline python/ribasim/tests/test_model.py:154]: Why twice?

**Review by evetion — CHANGES_REQUESTED** [review]

File: python/ribasim/tests/test_model.py (hunk 12 above)
Comment (SouthEndMusic) [reply python/ribasim/tests/test_model.py]: The added model is a subnetwork with `allocation_network_id = 1` (which will probably be changed in the future because that is reserved for the main network), so I check with `assert model.max_allocation_network_id() == 2` whether a shift was added to the `allocation_network_id` of the second one. I added some clarification in the code. I'll also test the coordinate translation while I'm at it.

File: python/ribasim/ribasim/model.py (hunk 11 above)
Comment (SouthEndMusic) [reply python/ribasim/ribasim/model.py:585]: You mean the whole subnetwork plotting thing? I guess it would have bean cleaner if that was done in a separate PR, but it did help with debugging of the model merging

File: python/ribasim/tests/test_model.py (hunk 1 above)
Comment (SouthEndMusic) [reply python/ribasim/tests/test_model.py:186]: https://github.com/Deltares/Ribasim/issues/928

**Comment by SouthEndMusic** [comment]
To do:
- [ ] Process comments by Maarten above
- [ ] Remove `Model.merge` options: `offset_node_id` and `offset_allocation_network_id` are always computed
- [ ] `offset_spatial`: make functions in tests
- [ ] `inplace`: don't make copy of entire model, only loop over nodes and modify those
- [ ] https://github.com/Deltares/Ribasim-NL/pull/49

File: python/ribasim/ribasim/input_base.py (hunk 5 above)
Comment (SouthEndMusic) [reply python/ribasim/ribasim/input_base.py]: I've added a check that the tables have no node IDs in common.

**Comment by evetion** [comment]
- [ ] Remove special code to merge models (default states, after its removed by another PR)
- [ ] Make merge methods on lower levels (node, tables) private
- [ ] Make use of model_copy(deep=True) and check whether it is needed everywhere
- [ ] Make inplace merge private (only merge (which copies) on model level)

**Comment by visr** [comment]
This is paused, after #1110 we can decide how to proceed with this.

**Comment by visr** [comment]
Closing this PR, since so many things changed that it probably needs a new start. The PR can still be used as inspiration for #912.
